packages/@aws-cdk/assets/lib/fs/copy.ts

```diff
@@ -2,56 +2,45 @@ import fs = require('fs');
 import path = require('path');
 import { CopyOptions } from './copy-options';
 import { FollowMode } from './follow-mode';
-import { shouldExclude, shouldFollow } from './utils';
+import { listFilesRecursively, shouldFollow } from './utils';
 
 export function copyDirectory(srcDir: string, destDir: string, options: CopyOptions = { }, rootDir?: string) {
   const follow = options.follow !== undefined ? options.follow : FollowMode.EXTERNAL;
-  const exclude = options.exclude || [];
 
   rootDir = rootDir || srcDir;
 
   if (!fs.statSync(srcDir).isDirectory()) {
     throw new Error(`${srcDir} is not a directory`);
   }
 
-  const files = fs.readdirSync(srcDir);
-  for (const file of files) {
-    const sourceFilePath = path.join(srcDir, file);
+  for (const sourceFilePath of listFilesRecursively(srcDir, {...options, follow}, rootDir)) {
+    const filePath = path.relative(rootDir, sourceFilePath);
+    const destFilePath = path.join(destDir, filePath);
 
-    if (shouldExclude(exclude, path.relative(rootDir, sourceFilePath))) {
-      continue;
-    }
-
-    const destFilePath = path.join(destDir, file);
-
-    let stat: fs.Stats | undefined = follow === FollowMode.ALWAYS
-      ? fs.statSync(sourceFilePath)
-      : fs.lstatSync(sourceFilePath);
-
-    if (stat && stat.isSymbolicLink()) {
-      const target = fs.readlinkSync(sourceFilePath);
+    if (follow !== FollowMode.ALWAYS) {
+      const stat = fs.lstatSync(sourceFilePath);
 
-      // determine if this is an external link (i.e. the target's absolute path
-      // is outside of the root directory).
-      const targetPath = path.normalize(path.resolve(srcDir, target));
+      if (stat && stat.isSymbolicLink()) {
+        const target = fs.readlinkSync(sourceFilePath);
+        const targetPath = path.normalize(path.resolve(srcDir, target));
+        if (!shouldFollow(follow, rootDir, targetPath)) {
+          fs.symlinkSync(target, destFilePath);
 
-      if (shouldFollow(follow, rootDir, targetPath)) {
-        stat = fs.statSync(sourceFilePath);
-      } else {
-        fs.symlinkSync(target, destFilePath);
-        stat = undefined;
+          continue;
+        }
       }
     }
 
-    if (stat && stat.isDirectory()) {
-      fs.mkdirSync(destFilePath);
-      copyDirectory(sourceFilePath, destFilePath, options, rootDir);
-      stat = undefined;
-    }
+    {
+      const destFileDir = path.dirname(destFilePath);
 
-    if (stat && stat.isFile()) {
-      fs.copyFileSync(sourceFilePath, destFilePath);
-      stat = undefined;
+      if (!fs.existsSync(destFileDir)) {
+        // FIXME fs.mkdirpSync(destFileDir);
```

Comment:
I've tried adding `fs-extra` as a depdency to `assets`:

```
lerna add fs-extra@^8.1.0 --scope=@aws-cdk/assets
lerna add --dev @types/fs-extra@^8.0.0 --scope=@aws-cdk/assets
```

But I get the following error when I try building the package:
`Error: Error: Unable to locate module: fs-extra`

Any idea?

Reply:
JSII modules cannot have non-JSII dependencies. You may only depend on JSII modules and the standard Node runtime.

Reply:
What should I do then? Add `mkdirpSync` to `@aws-cdk/core`?

Reply:
Reimplementing is the only viable solution.